rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1268 @@
+  - Otherwise, the body of `with attrs;` must start on a new line without any additional indentation.
+- If the body is `{ ... }`, `[ ... ]`, `( ... )` or `'' ... ''`
+  and `with attrs; …` is parenthesised or to the right side of a binding (e.g. `(with; [ …` or `foo = with; [ …`),
+  then `with` must start on the same line.
+  - Otherwise it must start on a new line.
+
+**Examples**
+
+```nix
+{
+  # Good
+  foo = with bar; [
+    # multiline
+    baz
+  ];
+  
+  # Good
+  foo =
+    with foo;
+    with bar; [
+      # multiline
+      baz
+    ];
+  
+  # Good
+  foo =
+    with bar;
+    baz foo {
+      # multiline
+      qux = 10;
+    };
+
+  # Good
+  foo =
+    with bar;
+    if cond then
+      foo
+    else
+      bar;
+  
+  # Bad
+  foo = assert qux; with bar; [
+    # multiline
+    baz
+  ];
+  
+  # Bad
+  foo = with bar;
+    [
+      # multiline
+      baz
+    ];
+
+  # Bad
+  foo =
+    with bar; [
+      # multiline
+      baz
+    ];
+    
+  # Good
+  [
+    qux
+    quux
+  ]
+  ++ (with pkgs; [
+    baz
+    blorp
+  ]);
+}
+```
+
+### let-in
+
+Let bindings must always have this form:
+```
+let
+  <name1> = <value1>;
+  <name2> = <value2>;
+  ...
+in
+<body>
+```
+
+- Let bindings are *always* multiline.
+- Each binding is indented and starts on its own line.
+  For more details, see the [bindings section](#bindings).
+- The `<body>` always starts on a new line and is not indented.
+
+**Examples**
+
+```nix
+let
+  foo = "bar";
+in
+func foo;
+
+let
+  foo = "bar";
+in
+{
+  inherit foo;
+  baz = "smth";
+}
+
+let
+  foo = "bar";
+in
+if foo == "bar" then
+  "hello"
+else
+  "world"
+```
+
+**Alternatives**
+
+- To allow having the `<body>` be on the same line as the `in`:
+  ```
+  let
+    <name1> = <value1>;
+    <name2> = <value2>;
+    ...
+  in <body>
+  ```
+
+  In particular when `<body>` is an identifier, list, attribute set and/or others.
+
+  Problems with this alternative:
+  - It leads to larger diffs when inserting something after the `in`
+  - The formatting can change when `<body>` is updated
+  - It's less consistent, since the formatting depends on the `<body>`
+
+- The body could be indented by a level
+  ```
+  let
+    <name1> = <value1>;
+    <name2> = <value2>;
+    ...
+  in
+    <body>
+  ```
+  
+  Problems with this alternative:
+  - Leads to indentation creeps
+  - Inconsistent with other expressions that have a `<body>` that is "returned"
+  - Favors a style where the body starts on the same line as the in for some values (e.g. attribute sets) to reduce an indentation level, see above.
+
+### Attribute sets and lists
+
+- Brackets and braces must always have a space (or line break) on the inside, like `[ `, ` ]`, `{ ` and ` }`.
+  - Empty lists and attribute sets are written as `[ ]` and `{ }`, respectively.
+- Lists and attribute sets with multiple items should be liberally expanded.
+  - They can only be on a single line if they fit on the line and contain few enough elements.
```

Comment:
Once you go through all relevant scenarios and edge cases, the rules specifying this may be pretty complex. It is not beneficial to nail them down in the RFC text with every detail. Instead, the text merely captures the *intent*, while leaving the actual implementation, well, as an implementation detail. As an example for edge cases, we currently have rules in Nixfmt that force-expand attribute sets in certain positions.

Generally, I like the format to mostly specify how syntax nodes look like in their expanded form (and their unexpanded form, but which is almost always trivial). On the other hand, the rules about *when* to expand something or not on are much less clearly defined.